File: src/decoders/discussion-messages.ts
import { DiscussionDraftMessage, DiscussionMessages, DiscussionSendAction, DiscussionSentMessage, SessionHandle } from "~/models";
import { decodeDiscussionDraftMessage } from "./discussion-draft-message";
import { decodeDiscussionSentMessage } from "./discussion-sent-message";

export const decodeDiscussionMessages = (messages: any, session: SessionHandle): DiscussionMessages => {
  const draft = messages.brouillon;
  const sents: DiscussionSentMessage[] = [];
  const drafts: DiscussionDraftMessage[] = [];
  const defaultReplyMessageID: string = messages.messagePourReponse.V.N;

  for (const message of messages.listeMessages.V) {
    if (message.brouillon) drafts.push(decodeDiscussionDraftMessage(message));
    else sents.push(decodeDiscussionSentMessage(message, session, sents));
  }

  sents.sort((a, b) => b.creationDate.getTime() - a.creationDate.getTime());

  if (draft) {
    const isHTML = draft.V.estHTML ?? false;

    drafts.push({
      isHTML,
      content: isHTML ? draft.V.contenu : draft.V.contenu.V,
      posessionID: draft.V.N,
      replyMessageID: defaultReplyMessageID
    });
  }

  const sendAction: DiscussionSendAction | undefined = messages.listeBoutons.V.find((button: any) => button.L.startsWith("Envoyer"))?.G;
  const canIncludeStudentsAndParents = sendAction === DiscussionSendAction.ReplyEveryoneExceptParentsAndStudents
                                    || sendAction === DiscussionSendAction.SendEveryoneExceptParentsAndStudents;

  return {
    sents,
    drafts,
    defaultReplyMessageID,
    sendAction,
    canIncludeStudentsAndParents
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/discussion-messages.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.*
import kotlinx.serialization.json.*

fun decodeDiscussionMessages (messages: JsonObject, session: SessionHandle): DiscussionMessages {
    val draft = messages["brouillon"]
    val sents: MutableList<DiscussionMessage<DiscussionSentMessage>> = mutableListOf()
    val drafts: MutableList<DiscussionDraftMessage> = mutableListOf()
    val defaultReplyMessageID = messages["messagePourReponse"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content

    for (message in messages["listeMessages"]!!.jsonObject["V"]!!.jsonArray)
        if (draft != null) drafts.add(decodeDiscussionDraftMessage(message.jsonObject))
        else sents.add(decodeDiscussionMessage(message.jsonObject, session, { decodeDiscussionSentMessage(message.jsonObject, session, sents) }))

    sents.sortedWith {a, b -> b.creationDate.time.toMillisecondOfDay() - a.creationDate.time.toMillisecondOfDay()}

    if (draft != null) {
        val isHTML = draft.jsonObject["V"]!!.jsonObject["estHTML"]?.jsonPrimitive?.boolean ?: false

        drafts.add(
            DiscussionDraftMessage(
                isHTML = isHTML,
                content = if (isHTML) draft.jsonObject["V"]!!.jsonObject["contenu"]!!.jsonPrimitive.content else draft.jsonObject["V"]!!.jsonObject["contenu"]!!.jsonObject["V"]!!.jsonPrimitive.content,
                possessionID = draft.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content,
                replyMessageID = defaultReplyMessageID
            )
        )
    }

    val actionObj = messages["listeBoutons"]!!.jsonObject["V"]!!.jsonArray.find { it.jsonObject["L"]!!.jsonPrimitive.content.startsWith("Envoyer") }
    val sendAction: DiscussionSendAction? = if (actionObj != null) DiscussionSendAction.fromInt(actionObj.jsonObject["G"]!!.jsonPrimitive.int) else null
    val canIncludeStudentsAndParents = sendAction == DiscussionSendAction.ReplyEveryoneExceptParentsAndStudents
            || sendAction == DiscussionSendAction.SendEveryoneExceptParentsAndStudents

    return DiscussionMessages(
        sents = sents,
        drafts = drafts,
        defaultReplyMessageID = defaultReplyMessageID,
        sendAction = sendAction,
        canIncludeStudentsAndParents = canIncludeStudentsAndParents
    )
}